elasticsearch_acls.go

```diff
@@ -0,0 +1,190 @@
+	ElasticSearchACLsHandler struct {
+		client *Client
+	}
+
+	// ElasticsearchACLRequest Aiven API request
+	// https://api.aiven.io/v1/project/<project>/service/<service_name>/elasticsearch/acl
+	ElasticsearchACLRequest struct {
+		ElasticSearchACLConfig ElasticSearchACLConfig `json:"elasticsearch_acl_config"`
+	}
+
+	// ElasticSearchACLResponse Aiven API response
+	// https://api.aiven.io/v1/project/<project>/service/<service_name>/elasticsearch/acl
+	ElasticSearchACLResponse struct {
+		APIResponse
+		ElasticSearchACLConfig ElasticSearchACLConfig `json:"elasticsearch_acl_config"`
+	}
+
+	// ElasticSearchACLConfig represents a configuration for Elasticsearch ACLs
+	ElasticSearchACLConfig struct {
+		ACLs        []ElasticSearchACL `json:"acls"`
+		Enabled     bool               `json:"enabled"`
+		ExtendedAcl bool               `json:"extendedAcl"`
+	}
+
+	// ElasticSearchACL represents a ElasticSearch ACLs entry
+	ElasticSearchACL struct {
+		Rules    []ElasticsearchACLRule `json:"rules"`
+		Username string                 `json:"username"`
+	}
+
+	// ElasticsearchACLRule represents a ElasticSearch ACLs Rule entry
+	ElasticsearchACLRule struct {
+		Index      string `json:"index"`
+		Permission string `json:"permission"`
+	}
+)
+
+// Create creates new Elasticsearch ACLs config
+func (h *ElasticSearchACLsHandler) Create(project, service string, req ElasticsearchACLRequest) (*ElasticSearchACLResponse, error) {
+	path := buildPath("project", project, "service", service, "elasticsearch", "acl")
+	bts, err := h.client.doPostRequest(path, req)
+	if err != nil {
+		return nil, err
+	}
+
+	return h.response(bts)
+}
+
+// Update updates Elasticsearch ACL config
+func (h *ElasticSearchACLsHandler) Update(project, service string, req ElasticsearchACLRequest) (*ElasticSearchACLResponse, error) {
+	path := buildPath("project", project, "service", service, "elasticsearch", "acl")
+	bts, err := h.client.doPutRequest(path, req)
+	if err != nil {
+		return nil, err
+	}
+
+	return h.response(bts)
+}
+
+// Get gets all existing Elasticsearch ACLs config
+func (h *ElasticSearchACLsHandler) Get(project, service string) (*ElasticSearchACLResponse, error) {
+	path := buildPath("project", project, "service", service, "elasticsearch", "acl")
+	bts, err := h.client.doGetRequest(path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return h.response(bts)
+}
+
+// Add function adds one new ACL record to already existing Elasticsearch ACL configuration,
+// it first checks if such ES user already exists. And if yes goes through the list of
+// ACL rules for such user and check if new rules that user tries to add already exists in
+// the configuration or not. We only add new records to the Elasticsearch configuration if case
+// there is a difference between actual existing configuration and what user is willing to add.
+func (h *ElasticSearchACLsHandler) Add(project, service string, acl ElasticSearchACL) error {
```

Comment:
Is this actually needed?

Reply:
I reckon yes, imagine a situation when we have a company which is actively using Elasticsearch ACLs. Lets they created a terraform script which generates user A with some permissions. Still, some other employees from this company go to the web interface and add new user B.

 If we did not provide Add method, they would need to fire Update with entire config which based on Terraform script only has user A, and when they run Terraform apply next time user B will be deleted from ACLs list.

Reply:
For example, Kafka ALC API endpoint works differently from Elasticsearch ACLs, when user whats to add new Kafka ACL they send a POST request to `Kafka/ACL` and new ACL gets appended to Kafka ACL list. But Elasticsearch ACLs endpoint work differently, the user always has to provide the entire list of ACLs and these new lists will replace the old one. API interface does not offer functionality to append or add new ACL to Elasticssearch ACLlist that is why I belave these methods are required.